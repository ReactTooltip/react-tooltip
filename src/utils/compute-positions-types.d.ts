import type { Middleware } from '../components/Tooltip/TooltipTypes'

export interface IComputePositions {
  elementReference?: Element | HTMLElement | null
  tooltipReference?: Element | HTMLElement | null
  tooltipArrowReference?: Element | HTMLElement | null
  place?: 'top' | 'right' | 'bottom' | 'left'
  offset?: number
  strategy?: 'absolute' | 'fixed'
  middlewares?: Middleware[]
}
