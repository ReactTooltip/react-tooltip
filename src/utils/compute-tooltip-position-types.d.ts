import type { CSSProperties } from 'react'
import type { Middleware, PlacesType } from '../components/Tooltip/TooltipTypes'

export interface IComputePositionArgs {
  elementReference?: Element | HTMLElement | null
  tooltipReference?: Element | HTMLElement | null
  tooltipArrowReference?: Element | HTMLElement | null
  place?: PlacesType
  offset?: number
  strategy?: 'absolute' | 'fixed'
  middlewares?: Middleware[]
  border?: CSSProperties['border']
}

export interface IComputedPosition {
  tooltipStyles: {
    left?: string
    top?: string
    border?: CSSProperties['border']
  }
  tooltipArrowStyles: {
    left?: string
    top?: string
    right?: string
    bottom?: string
    borderRight?: CSSProperties['border']
    borderBottom?: CSSProperties['border']
  }
  place: PlacesType
}
