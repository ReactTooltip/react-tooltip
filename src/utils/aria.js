/**
 * Support aria- and role in ReactTooltip
 *
 * @params props {Object}
 * @return {Object}
 */
export function parseAria (props) {
  let ariaObj = {}
  Object.keys(props).filter(prop => {
    // aria-xxx and role is acceptable
    return /(^aria-\w+$|^role$)/.test(prop)
  }).forEach(prop => {
    ariaObj[prop] = props[prop]
  })

  return ariaObj
}
