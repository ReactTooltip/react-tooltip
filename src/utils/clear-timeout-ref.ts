const clearTimeoutRef = (ref: React.MutableRefObject<NodeJS.Timeout | null>) => {
  if (ref.current) {
    clearTimeout(ref.current)
    // eslint-disable-next-line no-param-reassign
    ref.current = null
  }
}

export default clearTimeoutRef
