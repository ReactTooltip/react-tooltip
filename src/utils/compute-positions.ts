import { computePosition, offset, shift, arrow, flip } from '@floating-ui/dom'
import type { IComputePositions } from './compute-positions-types'

export const computeTooltipPosition = async ({
  elementReference = null,
  tooltipReference = null,
  tooltipArrowReference = null,
  place = 'top',
  offset: offsetValue = 10,
  strategy = 'absolute',
  middlewares = [offset(Number(offsetValue)), flip(), shift({ padding: 5 })],
}: IComputePositions) => {
  if (!elementReference) {
    // elementReference can be null or undefined and we will not compute the position
    // eslint-disable-next-line no-console
    // console.error('The reference element for tooltip was not defined: ', elementReference)
    return { tooltipStyles: {}, tooltipArrowStyles: {} }
  }

  if (tooltipReference === null) {
    return { tooltipStyles: {}, tooltipArrowStyles: {} }
  }

  const middleware = middlewares

  if (tooltipArrowReference) {
    middleware.push(arrow({ element: tooltipArrowReference as HTMLElement, padding: 5 }))

    return computePosition(elementReference as HTMLElement, tooltipReference as HTMLElement, {
      placement: place,
      strategy,
      middleware,
    }).then(({ x, y, placement, middlewareData }) => {
      const styles = { left: `${x}px`, top: `${y}px` }

      const { x: arrowX, y: arrowY } = middlewareData.arrow ?? { x: 0, y: 0 }

      const staticSide =
        {
          top: 'bottom',
          right: 'left',
          bottom: 'top',
          left: 'right',
        }[placement.split('-')[0]] ?? 'bottom'

      const arrowStyle = {
        left: arrowX != null ? `${arrowX}px` : '',
        top: arrowY != null ? `${arrowY}px` : '',
        right: '',
        bottom: '',
        [staticSide]: '-4px',
      }

      return { tooltipStyles: styles, tooltipArrowStyles: arrowStyle }
    })
  }

  return computePosition(elementReference as HTMLElement, tooltipReference as HTMLElement, {
    placement: 'bottom',
    strategy,
    middleware,
  }).then(({ x, y }) => {
    const styles = { left: `${x}px`, top: `${y}px` }

    return { tooltipStyles: styles, tooltipArrowStyles: {} }
  })
}
