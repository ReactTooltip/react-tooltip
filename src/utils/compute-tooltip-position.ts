import { computePosition, offset, shift, arrow, flip } from '@floating-ui/dom'
import type { IComputePositionArgs } from './compute-tooltip-position-types'

const computeTooltipPosition = async ({
  elementReference = null,
  tooltipReference = null,
  tooltipArrowReference = null,
  place = 'top',
  offset: offsetValue = 10,
  strategy = 'absolute',
  middlewares = [
    offset(Number(offsetValue)),
    flip({
      fallbackAxisSideDirection: 'start',
    }),
    shift({ padding: 5 }),
  ],
  border,
}: IComputePositionArgs) => {
  if (!elementReference) {
    // elementReference can be null or undefined and we will not compute the position
    // eslint-disable-next-line no-console
    // console.error('The reference element for tooltip was not defined: ', elementReference)
    return { tooltipStyles: {}, tooltipArrowStyles: {}, place }
  }

  if (tooltipReference === null) {
    return { tooltipStyles: {}, tooltipArrowStyles: {}, place }
  }

  const middleware = middlewares

  if (tooltipArrowReference) {
    middleware.push(arrow({ element: tooltipArrowReference as HTMLElement, padding: 5 }))

    return computePosition(elementReference as HTMLElement, tooltipReference as HTMLElement, {
      placement: place,
      strategy,
      middleware,
    }).then(({ x, y, placement, middlewareData }) => {
      const styles = { left: `${x}px`, top: `${y}px`, border }

      /* c8 ignore start */
      const { x: arrowX, y: arrowY } = middlewareData.arrow ?? { x: 0, y: 0 }

      const staticSide =
        {
          top: 'bottom',
          right: 'left',
          bottom: 'top',
          left: 'right',
        }[placement.split('-')[0]] ?? 'bottom'
      /* c8 ignore end */

      const borderSide = border && {
        borderBottom: border,
        borderRight: border,
      }

      let borderWidth = 0
      if (border) {
        const match = `${border}`.match(/(\d+)px/)
        if (match?.[1]) {
          borderWidth = Number(match[1])
        } else {
          /**
           * this means `border` was set without `width`,
           * or non-px value (such as `medium`, `thick`, ...)
           */
          borderWidth = 1
        }
      }

      /* c8 ignore start */
      const arrowStyle = {
        left: arrowX != null ? `${arrowX}px` : '',
        top: arrowY != null ? `${arrowY}px` : '',
        right: '',
        bottom: '',
        ...borderSide,
        [staticSide]: `-${4 + borderWidth}px`,
      }
      /* c8 ignore end */

      return {
        tooltipStyles: styles,
        tooltipArrowStyles: arrowStyle,
        place: placement,
      }
    })
  }

  return computePosition(elementReference as HTMLElement, tooltipReference as HTMLElement, {
    placement: 'bottom',
    strategy,
    middleware,
  }).then(({ x, y, placement }) => {
    const styles = { left: `${x}px`, top: `${y}px` }

    return { tooltipStyles: styles, tooltipArrowStyles: {}, place: placement }
  })
}

export default computeTooltipPosition
