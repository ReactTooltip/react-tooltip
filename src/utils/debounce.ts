/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * This function debounce the received function
 * @param { function } 	func				Function to be debounced
 * @param { number } 		wait				Time to wait before execut the function
 * @param { boolean } 	immediate		Param to define if the function will be executed immediately
 */
const debounce = (func: (...args: any[]) => void, wait?: number, immediate?: true) => {
  let timeout: NodeJS.Timeout | null = null

  return function debounced(this: typeof func, ...args: any[]) {
    const later = () => {
      timeout = null
      if (!immediate) {
        func.apply(this, args)
      }
    }

    if (immediate && !timeout) {
      /**
       * there's not need to clear the timeout
       * since we expect it to resolve and set `timeout = null`
       */
      func.apply(this, args)
      timeout = setTimeout(later, wait)
    }

    if (!immediate) {
      if (timeout) {
        clearTimeout(timeout)
      }
      timeout = setTimeout(later, wait)
    }
  }
}

export default debounce
