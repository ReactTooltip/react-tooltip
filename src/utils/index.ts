import type { IComputedPosition } from './compute-tooltip-position-types'
import computeTooltipPosition from './compute-tooltip-position'
import cssSupports from './css-supports'
import cssTimeToMs from './css-time-to-ms'
import debounce from './debounce'
import deepEqual from './deep-equal'
import getScrollParent from './get-scroll-parent'
import useIsomorphicLayoutEffect from './use-isomorphic-layout-effect'

export type { IComputedPosition }
export {
  computeTooltipPosition,
  cssSupports,
  cssTimeToMs,
  debounce,
  deepEqual,
  getScrollParent,
  useIsomorphicLayoutEffect,
}
