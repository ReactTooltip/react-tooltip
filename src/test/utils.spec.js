import debounce from 'utils/debounce'
import { computeTooltipPosition } from 'utils/compute-positions'

// Tell Jest to mock all timeout functions
jest.useRealTimers()

describe('compute positions', () => {
  test('empty reference elements', async () => {
    const value = await computeTooltipPosition({
      elementReference: null,
      tooltipReference: null,
      tooltipArrowReference: null,
    })

    expect(value).toEqual({ tooltipStyles: {}, tooltipArrowStyles: {} })
  })

  test('empty tooltip reference element', async () => {
    const element = document.createElement('div')
    const value = await computeTooltipPosition({
      elementReference: element,
      tooltipReference: null,
      tooltipArrowReference: null,
    })

    expect(value).toEqual({ tooltipStyles: {}, tooltipArrowStyles: {} })
  })

  test('empty tooltip arrow reference element', async () => {
    const element = document.createElement('div')
    const elementTooltip = document.createElement('div')
    const value = await computeTooltipPosition({
      elementReference: element,
      tooltipReference: elementTooltip,
      tooltipArrowReference: null,
    })

    expect(value).toEqual({
      tooltipArrowStyles: {},
      tooltipStyles: {
        left: '5px',
        top: '10px',
      },
    })
  })

  test('all reference elements', async () => {
    const element = document.createElement('div')
    const elementTooltip = document.createElement('div')
    const elementTooltipArrow = document.createElement('div')
    const value = await computeTooltipPosition({
      elementReference: element,
      tooltipReference: elementTooltip,
      tooltipArrowReference: elementTooltipArrow,
    })

    expect(value).toEqual({
      tooltipArrowStyles: {
        bottom: '-4px',
        left: '5px',
        right: '',
        top: '',
      },
      tooltipStyles: {
        left: '5px',
        top: '-10px',
      },
    })
  })
})

describe('debounce', () => {
  jest.useFakeTimers()

  let func
  let debouncedFunc

  beforeEach((timeout = 1000) => {
    func = jest.fn()
    debouncedFunc = debounce(func, timeout)
  })

  test('execute just once', () => {
    for (let i = 0; i < 100; i += 1) {
      debouncedFunc()
    }

    // Fast-forward time
    jest.runAllTimers()

    expect(func).toBeCalledTimes(1)
  })
})
