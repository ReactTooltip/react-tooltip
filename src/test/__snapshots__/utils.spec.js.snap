// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compute positions all reference elements 1`] = `
{
  "place": "top",
  "tooltipArrowStyles": {
    "bottom": "-4px",
    "left": "-1px",
    "right": "",
    "top": "",
  },
  "tooltipStyles": {
    "border": undefined,
    "left": "5px",
    "top": "-10px",
  },
}
`;
